### PR TITLE
chore(datadog_logs sink): Add log namespace support to `datadog_logs` sink

### PR DESCRIPTION
**Pull request by fuchsnj**
ref: https://github.com/vectordotdev/vector/issues/15453

This adds log namespace support to the datadog logs sink.
This is one of the few (maybe the only?) sink that previously supported semantic meaning, so those places were not touched. Everything else was migrated to new helper functions which should be used instead of accessing data using the "Global Log Schema".

These helper functions look at both the `log_namespace` and semantic meanings associated with an event to determine where messages, host, timestamp, source_type live.

### CODE REVIEW

File: src/sinks/datadog/logs/sink.rs [hunk 1]
```diff
@@ -134,9 +134,17 @@ impl crate::sinks::util::encoding::Encoder<Vec<Event>> for JsonEncoding {
...
             let log = event.as_mut_log();
-            log.rename_key(self.log_schema.message_key(), event_path!("message"));
-            log.rename_key(self.log_schema.host_key(), event_path!("host"));
-            if let Some(Value::Timestamp(ts)) = log.remove(self.log_schema.timestamp_key()) {
+            if let Some(message_path) = log.message_path() {
+                log.rename_key(message_path.as_str(), event_path!("message"));
+            }
+            if let Some(host_path) = log.host_path() {
+                log.rename_key(host_path.as_str(), event_path!("host"));
+            }
```
Comment (spencergilbert) [inline src/sinks/datadog/logs/sink.rs:142]: 🤔 don't we sort of need this to be infallible, payloads without these would be invalid. Shouldn't we handle that here?

Comment (fuchsnj) [reply src/sinks/datadog/logs/sink.rs:142]: The problem is that these can only return `None` in the case that semantic meaning isn't defined for these fields. "Sink Requirements" can enforce that semantic meaning is set correctly for sinks, but that feature isn't stable / is opt-in right now.

I guess I should just make these panic if it doesn't exist, to match the current behavior in this sink.

Comment (JeanMertz) [reply src/sinks/datadog/logs/sink.rs:142]: Well, that's awkward. Apparently I had this whole reply written, but never submitted the "review" 😅

File: lib/vector-core/src/event/log_event.rs [hunk 2]
```diff
@@ -399,6 +399,90 @@ impl LogEvent {
...
+    /// Fetches the `timestamp` of the event. This is either from the "timestamp" semantic meaning (Vector namespace)
+    /// or from the timestamp key set on the "Global Log Schema" (Legacy namespace).
+    pub fn get_timestamp(&self) -> Option<&Value> {
+        match self.namespace() {
+            LogNamespace::Vector => self.get_by_meaning("timestamp"),
+            LogNamespace::Legacy => self.get((PathPrefix::Event, log_schema().timestamp_key())),
+        }
+    }
+
+    /// Fetches the `timestamp` of the event. This is either from the "host" semantic meaning (Vector namespace)
```
Comment (spencergilbert) [inline lib/vector-core/src/event/log_event.rs:467]:
```suggestion
    /// Fetches the `host` of the event. This is either from the "host" semantic meaning (Vector namespace)
```

File: src/sinks/datadog/logs/sink.rs [hunk 3]
```diff
@@ -140,9 +140,20 @@ impl crate::sinks::util::encoding::Encoder<Vec<Event>> for JsonEncoding {
     fn encode_input(&self, mut input: Vec<Event>, writer: &mut dyn io::Write) -> io::Result<usize> {
         for event in input.iter_mut() {
             let log = event.as_mut_log();
-            log.rename_key(self.log_schema.message_key(), event_path!("message"));
-            log.rename_key(self.log_schema.host_key(), event_path!("host"));
-            if let Some(Value::Timestamp(ts)) = log.remove(self.log_schema.timestamp_key()) {
+            let message_path = log
+                .message_path()
+                .expect("message is required (make sure the \"message\" semantic meaning is set)");
```
Comment (spencergilbert) [inline src/sinks/datadog/logs/sink.rs:145]: Do we actually want to panic here, could there be times where the log doesn't have a "message" path?

Same for host.

Comment (fuchsnj) [reply src/sinks/datadog/logs/sink.rs:145]: This was changed to a panic based on the conversation here: https://github.com/vectordotdev/vector/pull/15473#discussion_r1041368995

Comment (spencergilbert) [reply src/sinks/datadog/logs/sink.rs:145]: Oh right, that sounds familiar now 😅

**Review by spencergilbert — APPROVED** [review]

**Review by StephenWakely — APPROVED** [review]

**Pull request merged**